README.md
# pycharm-llm-explain
This plugin uses the OpenAI API's GPT-4 turbo model to generate explanations for Python code in PyCharm Community Edition.

## Installation
(assuming you already cloned the repo)

### Sandboxed PyCharm
1. Open this repo in IntelliJ IDEA
2. Run the Gradle task "Run Plugin". This will open a sandboxes, fresh PyCharm instance with the plugin installed

## Running
1. Set the environment variable OPENAI_API_KEY to your OpenAI API Key
2. Select anywhere inside the method you wish to explain
3. Right click, then click "Show Context Actions", followed by "Explain code with OpenAI GPT-4"
4. Wait, and the explanation will appear in a popup!



CHANGELOG.md
<!-- Keep a Changelog guide -> https://keepachangelog.com -->

# pycharm-llm-explain Changelog

## [Unreleased]
### Added
- Initial scaffold created from [IntelliJ Platform Plugin Template](https://github.com/JetBrains/intellij-platform-plugin-template)


src/test/kotlin/com/github/gumjoe/pycharmllmexplain/MyPluginTest.kt
package com.github.gumjoe.pycharmllmexplain

import com.intellij.ide.highlighter.XmlFileType
import com.intellij.openapi.components.service
import com.intellij.psi.xml.XmlFile
import com.intellij.testFramework.TestDataPath
import com.intellij.testFramework.fixtures.BasePlatformTestCase
import com.intellij.util.PsiErrorElementUtil
import com.github.gumjoe.pycharmllmexplain.services.MyProjectService

@TestDataPath("\$CONTENT_ROOT/src/test/testData")
class MyPluginTest : BasePlatformTestCase() {

    fun testXMLFile() {
        val psiFile = myFixture.configureByText(XmlFileType.INSTANCE, "<foo>bar</foo>")
        val xmlFile = assertInstanceOf(psiFile, XmlFile::class.java)

        assertFalse(PsiErrorElementUtil.hasErrors(project, xmlFile.virtualFile))

        assertNotNull(xmlFile.rootTag)

        xmlFile.rootTag?.let {
            assertEquals("foo", it.name)
            assertEquals("bar", it.value.text)
        }
    }

    fun testRename() {
        myFixture.testRename("foo.xml", "foo_after.xml", "a2")
    }

    fun testProjectService() {
        val projectService = project.service<MyProjectService>()

        assertNotSame(projectService.getRandomNumber(), projectService.getRandomNumber())
    }

    override fun getTestDataPath() = "src/test/testData/rename"
}


src/main/kotlin/com/github/gumjoe/pycharmllmexplain/intentions/openai.kt
package com.github.gumjoe.pycharmllmexplain.intentions

import com.intellij.openapi.ui.Messages
import com.theokanning.openai.client.OpenAiApi
import com.theokanning.openai.service.OpenAiService
import okhttp3.OkHttpClient
import java.time.Duration

fun getOpenAi(): OpenAiService? {
    // Load OpenAI
    val openAiToken = System.getenv("OPENAI_API_KEY")
    if (openAiToken == null) {
        Messages.showErrorDialog(
            "Please specify your OpenAI API key in the environment variable OPENAI_API_KEY",
            "Plugin Error"
        )
        return null
    }

    // Code From https://github.com/TheoKanning/openai-java - needed to extend timeout
    val mapper = OpenAiService.defaultObjectMapper()
    val client: OkHttpClient = OpenAiService.defaultClient(openAiToken, Duration.ofMinutes(2))
        .newBuilder()
        .build()
    val retrofit = OpenAiService.defaultRetrofit(client, mapper)

    val api: OpenAiApi = retrofit.create(OpenAiApi::class.java)
    val service = OpenAiService(api)

    return service
}

src/main/kotlin/com/github/gumjoe/pycharmllmexplain/intentions/OpenAiExplain.kt
package com.github.gumjoe.pycharmllmexplain.intentions

import com.intellij.codeInsight.intention.IntentionAction
import com.intellij.codeInsight.intention.PsiElementBaseIntentionAction
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.progress.Task
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.psi.*
import com.intellij.psi.util.PsiTreeUtil
import com.jetbrains.python.psi.impl.PyFunctionImpl
import com.theokanning.openai.completion.chat.ChatCompletionRequest
import com.theokanning.openai.completion.chat.ChatMessage
import org.jetbrains.annotations.NonNls
import javax.swing.SwingUtilities

@NonNls
class OpenAiExplain : IntentionAction, PsiElementBaseIntentionAction() {

    private val log: Logger = Logger.getInstance(OpenAiExplain::class.java)

    override fun getText(): String {
        return "Explain code with OpenAI GPT-4"
    }

    override fun getFamilyName(): String {
        return "Documentation"
    }

    override fun startInWriteAction(): Boolean {
        return false
    }

    /**
     * Determines if we can explain the given element
     *
     * @param project the current project
     * @param editor the current editor (can be null)
     * @param element the element to check for a parent method
     * @return true if code can be explained, false otherwise
     */
    override fun isAvailable(project: Project, editor: Editor?, element: PsiElement): Boolean {
        val parent = PsiTreeUtil.getParentOfType(element, PyFunctionImpl::class.java)
        // Only trigger if method
        if (parent != null) {
            return true
        }
        return false
    }

    /**
     * Generates a summary of a python method
     *
     * @param project the current project
     * @param editor the current editor (can be null)
     * @param element the PsiElement representing the method to generate documentation for
     *
     * The method first grabs the source code of the given method. It then extracts all references to other methods, which are used as context for the OpenAI GPT-4 model.
     * The method then generates a prompt for the OpenAI GPT-4 model, which is then used to generate a summary of the method.
     *
     * If the combined method references + method body is too long we summarise each reference individually, and then summarise the method body using the summaries of the references as contetx instead of their raw source code
     *
     * If the OpenAI API key is not specified in the environment variable OPENAI_API_KEY, an error dialog is shown.
     * If OpenAI does not return a response or the response is invalid, an error dialog is shown.
     */
    override fun invoke(project: Project, editor: Editor?, element: PsiElement) {

        // Setup - grab method (which was already checked) and source code

        val method = PsiTreeUtil.getParentOfType(element, PyFunctionImpl::class.java)!!
        val sourceCode = method.text

        // Load openai - if not available, show error dialog
        val service = getOpenAi()
        if (service == null) {
            SwingUtilities.invokeLater {
                Messages.showErrorDialog(
                    "Failed to get OpenAI Service",
                    "Please specify your OpenAI API key in the environment variable OPENAI_API_KEY to use this plugin"
                )
            }
            return
        }

        // Get ready
        val processor = MethodProcessor(method)
        val preparedMethod = processor.prepareMethod()

        // Run our request to OpenAI in a progress manager
        ProgressManager.getInstance()
            .run(object : Task.Modal(project, "Requesting Explanation...", true) {
                override fun run(progressIndicator: ProgressIndicator) {
                    progressIndicator.isIndeterminate = true
                    progressIndicator.text = "Requesting explanation..."

                    log.debug("Requesting explanation from OpenAI...")
                    // Actually get the summary
                    var response = processor.runOpenAi(preparedMethod)
                    if (response == null) {
                        SwingUtilities.invokeLater {
                            Messages.showErrorDialog(
                                "OpenAI did not return a response!",
                                "OpenAI Error"
                            )
                        }
                        return
                    }

                    // If response enclosed within backticks, extract it!
                    if ("^\\s*```(.*)```\\s*$".toRegex(RegexOption.DOT_MATCHES_ALL) matches response) {
                        response = response.trim().drop(3).dropLast(3)
                    }

                    // DONE!
                    progressIndicator.fraction = 1.0

                    // Display result in dialog
                    SwingUtilities.invokeLater {
                        Messages.showInfoMessage(
                            "OpenAI response:\n$response",
                            "OpenAI Response"
                        )
                    }
                }
            })


        // Done!


    }

}

src/main/kotlin/com/github/gumjoe/pycharmllmexplain/intentions/MethodProcessor.kt
package com.github.gumjoe.pycharmllmexplain.intentions

import com.intellij.openapi.diagnostic.Logger
import com.intellij.psi.PsiElement
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.refactoring.suggested.startOffset
import com.jetbrains.python.psi.impl.PyCallExpressionImpl
import com.jetbrains.python.psi.impl.PyFunctionImpl
import com.knuddels.jtokkit.api.Encoding
import com.knuddels.jtokkit.api.EncodingRegistry
import com.knuddels.jtokkit.api.EncodingType
import com.theokanning.openai.completion.chat.ChatCompletionRequest
import com.theokanning.openai.completion.chat.ChatMessage


private val registry: EncodingRegistry = com.knuddels.jtokkit.Encodings.newDefaultEncodingRegistry()
private val enc: Encoding = registry.getEncoding(EncodingType.CL100K_BASE)
// We use gpt 4 turbo, which has 128k context
private const val MAX_CONTEXT_TOKENS = 100000

private const val OPENAI_SHORT_CHEAP_EXPLAINER_MODEL = "gpt-3.5-turbo"
private const val OPENAI_SUMMARY_MODEL = "gpt-4-1106-preview"
private const val OPENAI_TEMPERATURE = 0.1

class MethodProcessor(private val method: PyFunctionImpl) {

    private val log: Logger = Logger.getInstance(MethodProcessor::class.java)

    data class PreparedMethod(val name: String, val body: String, val references: Map<String, String>)

    /**
     * Returns a map of method references to their code
     */
    fun prepareMethod(): PreparedMethod {
        // 1: Extract method body
        val rootSourceCode = method.text
        // 2: Extract list of method refs inside method body
        val childMap = HashMap<String, PsiElement>()
        val children = PsiTreeUtil.findChildrenOfType(method, PyCallExpressionImpl::class.java)
        for (child in children) {
            val ref = child.callee?.reference?.resolve()
            if (ref != null) {
                val key = "${child.callee?.containingFile?.name}#${child.callee?.name}"
                log.debug("(${child.callee?.containingFile?.name}#${child.callee?.name}) - Offset: ${child.startOffset}, Offset in parent: ${child.startOffset - method.startOffset}")
                //log.warn(ref.text)
                if (childMap.containsKey(key)) {
                    log.warn("Duplicate key: $key")
                } else {
                    childMap[key] = ref
                }
            }
        }

        log.debug("============== DONE with ${method.name} ============")

        return PreparedMethod(method.name ?: "UNKNOWN", rootSourceCode, childMap.mapValues { (name, reference) -> reference.text })
    }

    public fun runOpenAi(method: PreparedMethod): String {
        // Step off point:
        // Conact method body and check length
        val minimumContent = StringBuilder()
        minimumContent.append(method.body)
        for (childBody in method.references.values) {
            minimumContent.append(childBody)
        }
        val tokens = enc.encode(minimumContent.toString())

        if (tokens.size > 0) {
            log.warn("Too many tokens!: ${tokens.size}")
        } else {
            log.debug("Tokens: ${tokens.size}")
        }

        // If too long, summarise the referenced methods and use that as context instead
        val referencesForSummary = if (tokens.size > MAX_CONTEXT_TOKENS) method.references.mapValues { (name, referenceBody) ->
            CodeReference(
                CodeReferenceTypes.OPENAI_SUMMARY, explainCode(
                    name,
                    referenceBody
                ), name
            )
        } else method.references.mapValues { (name, referenceBody) ->
            CodeReference(
                CodeReferenceTypes.RAW_CODE,
                referenceBody,
                name
            )
        }

        // NOTE: A future improvement may be to split the code up into chunks and summarise each chunk individually, then combine the summaries into a single summary, if we still have too much context

        // We can now create a summary of the method
        // 3: Generate summary of method body
        val summary = generateSummary(method.name, method.body, referencesForSummary)

        // Return it
        log.debug("Summary: $summary")
        return summary
    }

    private fun generateSummary(name: String?, rootSourceCode: String, referencesForSummary: Map<String, CodeReference>): String {
        val prompt = getSummaryPrompt(name ?: "UNKNOWN", rootSourceCode, referencesForSummary)
        log.debug("Prompt: $prompt")
        val completionRequest = ChatCompletionRequest.builder()
            .temperature(OPENAI_TEMPERATURE)
            .model(OPENAI_SUMMARY_MODEL)
            .messages(
                listOf(
                    ChatMessage("user", prompt)
                )
            )
            .build()
        log.debug("Requesting summary from OpenAI...")
        val service = getOpenAi()
        if (service == null) {
            log.warn("OpenAI service is null!")
            throw Exception("OpenAI service is null! Cannot proceed with processing.")
        }
        // Actually get the documentation
        val response =
            service.createChatCompletion(completionRequest).choices[0]?.message?.content?.toString()
        if (response == null) {
            log.warn("OpenAI response is null!")
            throw Exception("OpenAI response is null! Cannot proceed with processing.")
        }

        return response
    }

    private fun getSummaryPrompt(name: String, rootSourceCode: String, referencesForSummary: Map<String, CodeReference>): String? {
        val promptBuilder = StringBuilder()
        promptBuilder.append("You are ExplainPythonGPT, an expert at explaining Python code, even if the code is incomplete.\n")
        promptBuilder.append("You will first be provided with the method the code to explain references (either as code or a summary of the code), then the code itself to explain.\n")
        promptBuilder.append("\n")
        promptBuilder.append("Explain ONLY the code we ask to explain (denoted as \"Code to Explain:\"), noting however to always be honest with the user: if you do not know what part of a method does, say this.\n")
        promptBuilder.append("Explanations should be clear and concise and targeted at professional coders who want to quickly know what the code does.\n")
        promptBuilder.append("Not every line needs explaining in detail, we are interested more in what a method does and how it achieves that.\n")
        promptBuilder.append("Remember, if the method is extremely long you will need to be succient in your explanation as you only have so much output tokens available!\n")
        promptBuilder.append("\n")
        promptBuilder.append("## Method Metadata:\n")
        promptBuilder.append("Method Name: $name\n")
        promptBuilder.append("Number of references: ${referencesForSummary.size}\n")

        promptBuilder.append("## References\n")
        promptBuilder.append("\n")

        // Add references
        for (reference in referencesForSummary.values) {
            promptBuilder.append(referenceToPrompt(reference))
        }

        promptBuilder.append("\n## Code to Explain:\n")
        promptBuilder.append("```python\n")
        promptBuilder.append(rootSourceCode)
        promptBuilder.append("\n```\n")

        promptBuilder.append("Please provide the explanation of ONLY the Code to Explain.\n")
        return promptBuilder.toString()
    }

    private fun explainCode(name: String, sourceCode: String): String {
        log.debug("Explain: $name with GPT 3.5")
        val prompt = getExplainerPrompt(name, sourceCode)
        log.debug("Prompt: $prompt")
        val completionRequest = ChatCompletionRequest.builder()
            .temperature(OPENAI_TEMPERATURE)
            .model(OPENAI_SHORT_CHEAP_EXPLAINER_MODEL)
            .messages(
                listOf(
                    ChatMessage("user", prompt)
                )
            )
            .build()
        log.debug("Requesting explanation from OpenAI...")
        val service = getOpenAi()
        if (service == null) {
            log.warn("OpenAI service is null!")
            throw Exception("OpenAI service is null! Cannot proceed with processing.")
        }
        // Actually get the documentation
        val response =
            service.createChatCompletion(completionRequest).choices[0]?.message?.content?.toString()
        if (response == null) {
            log.warn("OpenAI response is null!")
            throw Exception("OpenAI response is null! Cannot proceed with processing.")
        }

        return response
    }

    private fun getExplainerPrompt(name: String, sourceCode: String): String {
        val prompt = StringBuilder()
        prompt.append("You are SummaryExplainGPT, an excellent code summarizer of Python code. Output ONLY short, 1-3 line summary of the code provided. It is important you keep the description short and succinct, i.e. a summary.\n")
        prompt.append("#### Name:\n")
        prompt.append(name + "\n")
        prompt.append("#### Source code:\n")
        prompt.append("```python\n")
        prompt.append(sourceCode)
        prompt.append("\n```\n")

        return prompt.toString()
    }
}

src/main/kotlin/com/github/gumjoe/pycharmllmexplain/toolWindow/MyToolWindowFactory.kt
package com.github.gumjoe.pycharmllmexplain.toolWindow

import com.intellij.openapi.components.service
import com.intellij.openapi.diagnostic.thisLogger
import com.intellij.openapi.project.Project
import com.intellij.openapi.wm.ToolWindow
import com.intellij.openapi.wm.ToolWindowFactory
import com.intellij.ui.components.JBLabel
import com.intellij.ui.components.JBPanel
import com.intellij.ui.content.ContentFactory
import com.github.gumjoe.pycharmllmexplain.MyBundle
import com.github.gumjoe.pycharmllmexplain.services.MyProjectService
import javax.swing.JButton


class MyToolWindowFactory : ToolWindowFactory {

    init {
        thisLogger().warn("Don't forget to remove all non-needed sample code files with their corresponding registration entries in `plugin.xml`.")
    }

    override fun createToolWindowContent(project: Project, toolWindow: ToolWindow) {
        val myToolWindow = MyToolWindow(toolWindow)
        val content = ContentFactory.getInstance().createContent(myToolWindow.getContent(), null, false)
        toolWindow.contentManager.addContent(content)
    }

    override fun shouldBeAvailable(project: Project) = true

    class MyToolWindow(toolWindow: ToolWindow) {

        private val service = toolWindow.project.service<MyProjectService>()

        fun getContent() = JBPanel<JBPanel<*>>().apply {
            val label = JBLabel(MyBundle.message("randomLabel", "?"))

            add(label)
            add(JButton(MyBundle.message("shuffle")).apply {
                addActionListener {
                    label.text = MyBundle.message("randomLabel", service.getRandomNumber())
                }
            })
        }
    }
}


src/main/kotlin/com/github/gumjoe/pycharmllmexplain/MyBundle.kt
package com.github.gumjoe.pycharmllmexplain

import com.intellij.codeInsight.intention.IntentionAction
import com.intellij.codeInsight.intention.PsiElementBaseIntentionAction
import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.progress.Task
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.psi.*
import com.intellij.psi.util.PsiTreeUtil
import com.jetbrains.python.codeInsight.override.PyMethodMember
import com.jetbrains.python.psi.impl.PyBaseElementImpl
import com.jetbrains.python.psi.impl.PyFunctionImpl
import org.intellij.plugins.relaxNG.compact.psi.util.PsiFunction
import org.jetbrains.annotations.NonNls
import javax.swing.SwingUtilities

const val OPENAI_MODEL = "gpt-3.5-turbo"
const val OPENAI_TEMPERATURE = 0.1

@NonNls
class OpenAIDoc : IntentionAction, PsiElementBaseIntentionAction() {

    private val log: Logger = Logger.getInstance(OpenAIDoc::class.java)

    override fun getText(): String {
        return "Explain code with OpenAI GPT-4"
    }

    override fun getFamilyName(): String {
        return "Documentation"
    }

    override fun startInWriteAction(): Boolean {
        return false
    }

    /**
     * Determines if we can explain the given element
     *
     * @param project the current project
     * @param editor the current editor (can be null)
     * @param element the element to check for a parent method
     * @return true if code can be explained, false otherwise
     */
    override fun isAvailable(project: Project, editor: Editor?, element: PsiElement): Boolean {
        val parent = PsiTreeUtil.getParentOfType(element, PyFunctionImpl::class.java)
        log.info(element::class.java.name)
        if (parent != null) {
            log.info(parent::class.java.name)
        }
        // First, is the parent a method?
//        if (parent != null && (parent is PsiFunction) {
//            // Made it
//            return true
//        }
//        return false
        return true
    }

    /**
     * Generates JavaDoc documentation for a given method using OpenAI's ChatGPT model.
     *
     * @param project the current project
     * @param editor the current editor (can be null)
     * @param element the PsiElement representing the method to generate documentation for
     *
     * The method first grabs the source code of the given method and the parent class's name and documentation comment.
     * It then assembles a prompt using this information and sends it to OpenAI's ChatGPT model to generate documentation.
     * The generated documentation is then validated and added to the method as a JavaDoc comment.
     *
     * If the OpenAI API key is not specified in the environment variable OPENAI_API_KEY, an error dialog is shown.
     * If OpenAI does not return a response or the response is invalid, an error dialog is shown.
     */
    override fun invoke(project: Project, editor: Editor?, element: PsiElement) {

        Messages.showErrorDialog(
            "Please specify your OpenAI API key in the environment variable OPENAI_API_KEY",
            "Plugin Error"
        )
        return
        // Setup - grab method (which was already checked) and source code

//        val method = atMethodContainer(element.parent)
//        val sourceCode = method.text
//
//        // Grab parent for OpenAI to use as context!
//        val parent = method.getContainingClass()
//        val parentComment = parent?.docComment?.text ?: "None provided"
//        val parentName = parent?.name ?: "None provided"
//
//        // Assemble prompt
//        val prompt = getPrompt(parentName, parentComment, sourceCode)
//
//        // Load OpenAI
//        val openAiToken = System.getenv("OPENAI_API_KEY")
//        if (openAiToken == null) {
//            Messages.showErrorDialog(
//                "Please specify your OpenAI API key in the environment variable OPENAI_API_KEY",
//                "Plugin Error"
//            )
//            return
//        }
//        val service = OpenAiService(openAiToken)
//
//        // Create request to OpenAI. We use ChatGPT as our responder.
//        val completionRequest = ChatCompletionRequest.builder()
//            .temperature(OPENAI_TEMPERATURE)
//            .model(OPENAI_MODEL)
//            .messages(
//                listOf(
//                    ChatMessage("user", prompt)
//                )
//            )
//            .build()
//
//        // Run our request to OpenAI in a progress manager
//        ProgressManager.getInstance()
//            .run(object : Task.Backgroundable(project, "Generating documentation", true) {
//                override fun run(progressIndicator: ProgressIndicator) {
//                    progressIndicator.isIndeterminate = true
//                    progressIndicator.text = "Generating documentation..."
//
//                    log.debug("Requesting documentation completion from OpenAI...")
//                    // Actually get the documentation
//                    var response =
//                        service.createChatCompletion(completionRequest).choices[0]?.message?.content?.toString()
//                    if (response == null) {
//                        SwingUtilities.invokeLater {
//                            Messages.showErrorDialog(
//                                "OpenAI did not return a response!",
//                                "OpenAI Error"
//                            )
//                        }
//                        return
//                    }
//
//                    // If response enclosed within backticks, extract it!
//                    if ("^\\s*```(.*)```\\s*$".toRegex(RegexOption.DOT_MATCHES_ALL) matches response) {
//                        response = response.trim().drop(3).dropLast(3)
//                    }
//
//                    // validate - did we actually get valid JavaDoc?
//                    if (!("^\\s*/\\*\\*(.*)\\*/\\s*$".toRegex(RegexOption.DOT_MATCHES_ALL) matches response)) {
//                        SwingUtilities.invokeLater {
//                            Messages.showErrorDialog(
//                                "OpenAI response was invalid. Got back $response",
//                                "OpenAI Error"
//                            )
//                        }
//                        return
//                    }
//
//                    // Have to wrap like this so that we can actually write to the source code.
//                    WriteCommandAction.runWriteCommandAction(project) {
//                        val factory = JavaPsiFacade.getElementFactory(project)
//                        val docComment = factory.createDocCommentFromText(response)
//                        method.addBefore(
//                            docComment,
//                            method.firstChild
//                        )
//                    }
//
//                    // DONE!
//                    progressIndicator.fraction = 1.0
//                }
//            })


        // Done!


    }

    /**
     * The actual prompt we use for ChatGPT
     */
    private fun getPrompt(parentClass: String, classComment: String, code: String) = run {
        """Output in the following format: A valid, informative, factual JavaDoc comment only. Do not enclose output with backticks or place output into a codeblock. Output ONLY the documentation (do not output or repeat back any code given to you).

Enclosing Class: $parentClass
Enclosing Class JavaDoc: ```
$classComment
```
Method to generate JavaDoc for:
```
$code
```
"""
    }

}